### ISSUE TITLE
Windows Inspec Train powershell error

### ISSUE THREAD

**Issue by Stromweld** [inspec/train#475]
# Version:
Chef Infra Client 15.1.36 with Inspec 4.6.4

# Environment:

Windows 2016, Chef Infra Client 15.1.36, audit 8.0.0 cookbook

# Scenario:

End of chef-client runs audit cookbook runs inspec profiles and is throwing an error

# Steps to Reproduce:
run `inspec init profile test` and then `inspec exec test`

# Expected Result:

Inspec runs the basic profile 

# Actual Result:
```
PS C:\opscode\chef\embedded\bin> .\inspec exec .\test
Traceback (most recent call last):
        19: from C:/opscode/chef/embedded/bin/inspec:23:in `<main>'
        18: from C:/opscode/chef/embedded/bin/inspec:23:in `load'
        17: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/inspec-core-bin-4.6.4/bin/inspec:11:in `<top (required)>'
        16: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
        15: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
        14: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
        13: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
        12: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/inspec-core-4.6.4/lib/inspec/cli.rb:294:in `exec'
        11: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/inspec-core-4.6.4/lib/inspec/cli.rb:294:in `new'
        10: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/inspec-core-4.6.4/lib/inspec/runner.rb:70:in `initialize'
         9: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/inspec-core-4.6.4/lib/inspec/runner.rb:78:in `configure_transport'
         8: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/inspec-core-4.6.4/lib/inspec/backend.rb:51:in `create'
         7: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/train-core-2.1.13/lib/train/transports/local.rb:17:in `connection'
         6: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/train-core-2.1.13/lib/train/transports/local.rb:17:in `new'
         5: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/train-core-2.1.13/lib/train/transports/local.rb:27:in `initialize'
         4: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/train-core-2.1.13/lib/train/transports/local.rb:52:in `select_runner'
         3: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/train-core-2.1.13/lib/train/transports/local.rb:52:in `new'
         2: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/train-core-2.1.13/lib/train/transports/local.rb:145:in `initialize'
         1: from C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/train-core-2.1.13/lib/train/transports/local.rb:163:in `acquire_pipe'
C:/opscode/chef/embedded/lib/ruby/gems/2.6.0/gems/train-core-2.1.13/lib/train/transports/local.rb:220:in `start_pipe_server': undefined method `process_id' for #<Array:0x00000000069c23f8> (NoMethodError)
```

**Comment by Stromweld** [comment]
@jerryaldrichiii asked me to submit this issue after a few debug steps we took in a slack chat https://chefcommunity.slack.com/archives/C1XB6U6MN/p1562701402176200

**Comment by Stromweld** [comment]
@tas50 just so your aware of this.

**Comment by Stromweld** [comment]
This appears to be having a larger affect on our servers. Each time the audit cookbook tries to execute inspec with it's profiles and the error above happens it's leaving behind a powershell process and they are accumulating a ridiculous amount of them. 
![Screen Shot 2019-07-26 at 12 33 08 PM](https://user-images.githubusercontent.com/3476391/61972308-e7ce4e00-afa6-11e9-8b89-7112de8ac1a1.png)
Here's what we also found in the logs for the process ID's:
```
07/19/2019 09:05:57 PM
LogName=Security
SourceName=Microsoft Windows security auditing.
EventCode=4688
EventType=0
Type=Information
ComputerName=ngsv4926.mstiller.com
TaskCategory=Process Creation
OpCode=Info
RecordNumber=5216709
Keywords=Audit Success
Message=A new process has been created.
 
Creator Subject:
            Security ID:                 NT AUTHORITY\SYSTEM
            Account Name:                       ngsv4926$
            Account Domain:                    MSTILLER
            Logon ID:                    0x3E7
 
Target Subject:
            Security ID:                 NULL SID
            Account Name:                       -
            Account Domain:                    -
            Logon ID:                    0x0
 
Process Information:
            New Process ID:                     0x2918
            New Process Name:   C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
            Token Elevation Type:            %%1936
            Mandatory Label:                    Mandatory Label\System Mandatory Level
            Creator Process ID:    0x2be0
            Creator Process Name:          C:\opscode\chef\embedded\bin\ruby.exe
            Process Command Line:       
__________________________________________________________________________________________
 
07/19/2019 09:36:01 PM
LogName=Security
SourceName=Microsoft Windows security auditing.
EventCode=4688
EventType=0
Type=Information
ComputerName=ngsv4926.mstiller.com
TaskCategory=Process Creation
OpCode=Info
RecordNumber=5217479
Keywords=Audit Success
Message=A new process has been created.
 
Creator Subject:
            Security ID:                 NT AUTHORITY\SYSTEM
            Account Name:                       ngsv4926$
            Account Domain:                    MSTILLER
            Logon ID:                    0x3E7
 
Target Subject:
            Security ID:                 NULL SID
            Account Name:                       -
            Account Domain:                    -
            Logon ID:                    0x0
 
Process Information:
            New Process ID:                     0x1920
            New Process Name:   C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
            Token Elevation Type:            %%1936
            Mandatory Label:                    Mandatory Label\System Mandatory Level
            Creator Process ID:    0x29f4
            Creator Process Name:          C:\opscode\chef\embedded\bin\ruby.exe
            Process Command Line:       
 
______________________________________________________________________________
 
07/19/2019 10:07:49 PM
LogName=Security
SourceName=Microsoft Windows security auditing.
EventCode=4688
EventType=0
Type=Information
ComputerName=ngsv4926.mstiller.com
TaskCategory=Process Creation
OpCode=Info
RecordNumber=5217992
Keywords=Audit Success
Message=A new process has been created.
 
Creator Subject:
            Security ID:                 NT AUTHORITY\SYSTEM
            Account Name:                       ngsv4926$
            Account Domain:                    MSTILLER
            Logon ID:                    0x3E7
 
Target Subject:
            Security ID:                 NULL SID
            Account Name:                       -
            Account Domain:                    -
            Logon ID:                    0x0
 
Process Information:
            New Process ID:                     0x2868
            New Process Name:   C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
            Token Elevation Type:            %%1936
            Mandatory Label:                    Mandatory Label\System Mandatory Level
            Creator Process ID:    0x24e0
            Creator Process Name:          C:\opscode\chef\embedded\bin\ruby.exe
            Process Command Line:
```

**Comment by jerryaldrichiii** [comment]
I suspect the issue here is due to how I had to rewrite how execution works remotely on Windows...

https://github.com/inspec/train/pull/220

TL;DR: Train used to spawn a new process for each PowerShell command. This took forever and the only solution I found was to used named_pipes, I suspect that each these are not being cleaned up correctly.

https://github.com/inspec/train/pull/220/files#diff-aaaf34d9f275e64de4830cd33fd6ed2cR185

**Comment by Stromweld** [comment]
That makes sense since it's not exiting properly and blowing up the processes aren't getting cleaned up.

**Comment by clintoncwolfe** [comment]
https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipes

https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipe-operations

**Comment by clintoncwolfe** [comment]
Currently teardown is attempted with an `at_exit` hook and a process kill - no explicit pipe close as recommended [here](https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipe-operations) (`DisconnectNamedPipe` and `CloseHandle`)

https://github.com/inspec/train/blob/e0960cd84acbdc150aa4df33d46e7064b48a586b/lib/train/transports/local.rb#L223

**Comment by zenspider** [comment]
This doesn't reproduce with the steps above.

**Comment by Stromweld** [comment]
@zenspider what version of chefdk and Inspec did you run? I’m running chef infra client 15.1.36 which bundles Inspec 4.6.4.

**Comment by zenspider** [comment]
Latest version of `chefdk` available on chocolatey, which apparently is different enough. 🤷‍♀ 

I now have a repro (and diagnosis (and fix?)) after installing chef-client via MSI. PR incoming (just need to resolve some perms so I can).

**Comment by zenspider** [comment]
This just got released as v3.0.7. It is being packaged into chefdk / chef-client. A release of inspec will also be going out shortly as well (but updating ANY of these packages should cover your bases).

**Comment by clintoncwolfe** [comment]
This has been fixed in #660
